sys/include/shell.h

```diff
@@ -48,6 +48,33 @@ extern "C" {
  */
 #define SHELL_DEFAULT_BUFSIZE   (128)
 
+/**
+ * @brief           Optional hook after readline has triggered.
+ * @details         User implemented function gets called after the shell
+ *                  readline is complete.
+ */
+void post_readline_hook(void);
+
+/**
+ * @brief           Optional hook before shell command is called.
+ * @details         User implemented function gets called before a valid shell
+ *                  command will be called.
+ *
+ * @param[in]       argc   Number of arguments supplied to the function invocation.
+ * @param[in]       argv   The supplied argument list.
+ */
+void pre_command_hook(int argc, char **argv);
+
+/**
+ * @brief           Optional hook after shell command is called.
+ * @details         User implemented function gets called before a valid shell
+ *                  command will be called.
+ *
+ * @param[in]       argc   Number of arguments supplied to the function invocation.
+ * @param[in]       argv   The supplied argument list.
+ */
+void post_command_hook(int argc, char **argv);
```

Comment:
In line with https://github.com/RIOT-OS/RIOT/pull/11732 we could factor in the result of the command here.

```suggestion
void post_command_hook(int res, int argc, char **argv);
```

Reply:
done